derive/src/decode.rs

```diff
@@ -300,3 +285,48 @@ fn create_instance(
 		},
 	}
 }
+
+pub fn quote_decode_with_mem_tracking_checks(data: &Data, crate_path: &syn::Path) -> TokenStream {
+	let fields: Box<dyn Iterator<Item = &Field>> = match data {
+		Data::Struct(data) => Box::new(data.fields.iter()),
+		Data::Enum(ref data) => {
+			let variants = match utils::try_get_variants(data) {
+				Ok(variants) => variants,
+				Err(e) => return e.to_compile_error(),
+			};
+
+			let mut fields: Box<dyn Iterator<Item = &Field>> = Box::new(iter::empty());
+			for variant in variants {
+				fields = Box::new(fields.chain(variant.fields.iter()));
+			}
+			fields
+		},
+		Data::Union(_) => {
+			return Error::new(Span::call_site(), "Union types are not supported.")
+				.to_compile_error();
+		},
+	};
+
+	let processed_fields = fields.filter_map(|field| {
+		if utils::should_skip(&field.attrs) {
+			return None;
+		}
+
+		let field_type = if let Some(compact) = utils::get_compact_type(field, crate_path) {
+			compact
+		} else if let Some(encoded_as) = utils::get_encoded_as_type(field) {
+			encoded_as
+		} else {
+			field.ty.to_token_stream()
+		};
+		Some(quote! {#field_type})
+	});
+
+	quote! {
```

Comment:
`quote_spanned` with the span of each type to get a better error. Maybe also a ui test would be nice

Reply:
Done. I hope I understood correctly how `quote_spanned` works.